presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -676,11 +819,45 @@ public void abort()
         return future;
     }
 
+    public static StreamingPlanSection extractStreamingSections(SubPlan subPlan)
+    {
+        List<SubPlan> childrenRoots = new ArrayList<>();
+        SubPlan streamingSection = extractStreamingSection(subPlan, childrenRoots);
+        return new StreamingPlanSection(
+                streamingSection,
+                childrenRoots.stream()
+                        .map(SqlQueryScheduler::extractStreamingSections)
+                        .collect(toImmutableList()));
+    }
+
+    private static SubPlan extractStreamingSection(SubPlan subPlan, List<SubPlan> otherRoots)
+    {
+        ImmutableList.Builder<SubPlan> streamingSources = ImmutableList.builder();
+        Set<PlanFragmentId> streamingFragmentIds = subPlan.getFragment().getRemoteSourceNodes().stream()
+                .map(RemoteSourceNode::getSourceFragmentIds)
+                .flatMap(List::stream)
+                .collect(toImmutableSet());
+        for (SubPlan child : subPlan.getChildren()) {
+            if (streamingFragmentIds.contains(child.getFragment().getId())) {
+                streamingSources.add(extractStreamingSection(child, otherRoots));
+            }
+            else {
+                otherRoots.add(child);
+            }
+        }
+        return new SubPlan(subPlan.getFragment(), streamingSources.build());
```

Comment:
This depends on the way you are thinking about `SubPlan`. Subplan has references to children, no mater if the children requires streaming or materializing exchange. This is only responsibility of a Scheduler to link them in a correct way. That's why i didn't introduce `StreamingSection` as a standalone class, but rather as a inner class of a `SqlQueryScheduler`. Technically nobody, but `SqlQueryScheduler` should now the details about the type of linkage between the subplans.

Reply:
> wow, this looks dangerous :/

I don't think that having 2 lists will make it any less dangerous. Is just that instead of this check: https://github.com/prestodb/presto/pull/12604/commits/fea302ad1d4f91c2983b26b5e50cfe1b437cb7da#diff-5cffc1bea86a011bf263a6d6a736c93fR507 you will have something like `checkArgument(subPlan.materializingChildren.isEmpty())`. It is still possible to overlook the `materializing` childrent altogether (or vice versa).

Reply:
Based on the discussion offline introduced `StreamingSubPlan` inner class in the `SqlQueryScheduler`